### PR TITLE
Enhance MicrosoftAuthClient to support binding between brokers

### PR DESCRIPTION
**Pull request by rpdome**
Check into dev for now. 

Will cherry-pick (and update .aidl order) if we're going to release this fix sooner.

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java [hunk 1]
```diff
@@ -140,4 +165,39 @@ private String getCurrentActiveBrokerPackageName(@NonNull final Context context)
...
+
+    /**
+     * Returns the inactive installed authenticator.
+     * - If CP is the broker, this will return authenticator.
+     * - If authenticator is the broker, this will return CP.
+     *
+     * @param context
+     * @return package name of the installed inactive broker app, if there's any.
+     */
+    private String getInactiveBrokerPackageName(@NonNull final Context context) {
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java]: Minor nit: can declare method `@Nullable`

File: common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java [hunk 2]
```diff
@@ -140,4 +165,39 @@ private String getCurrentActiveBrokerPackageName(@NonNull final Context context)
...
+        final String methodName = ":getInactiveBrokerPackageName";
+        String currentBroker = getCurrentActiveBrokerPackageName(context);
+
+        final String inactiveBrokerPackageName;
+        if (currentBroker.equalsIgnoreCase(AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_PACKAGE_NAME)) {
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java]: This looks to be a possible NPE... `getCurrentActiveBrokerPackageName()` will return `null` if there is no active broker -- flipping this check around a comparing against the constant would avoid this....

so

```java
if (AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_PACKAGE_NAME.equalsIgnoreCase(currentBroker)) {
    // ...
}
```

The same would apply to the `else if` clauses

File: common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java [hunk 3]
```diff
@@ -140,4 +166,40 @@ private String getCurrentActiveBrokerPackageName(@NonNull final Context context)
...
+        } else {
+            return null;
+        }
+
+        // Verify the signature to make sure that we're not binding to malicious apps.
+        final BrokerValidator validator = new BrokerValidator(context);
+        try {
+            if (validator.verifySignature(inactiveBrokerPackageName)) {
+                return inactiveBrokerPackageName;
+            }
```
Comment (kreedula) [inline common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java]: May be we should log here if verification fails

Comment (sdabbiru) [reply common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java]: Yes, actually, we should send telemetry if the signature check fails.

---
In reply to: [299733100](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/538#discussion_r299733100) [](ancestors = 299733100)

File: common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java [hunk 4]
```diff
@@ -140,4 +165,39 @@ private String getCurrentActiveBrokerPackageName(@NonNull final Context context)
...
+        } else if (currentBroker.equalsIgnoreCase(AuthenticationConstants.Broker.COMPANY_PORTAL_APP_PACKAGE_NAME)) {
+            inactiveBrokerPackageName = AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_PACKAGE_NAME;
+        } else {
+            return null;
+        }
+
+        // Verify the signature to make sure that we're not binding to malicious apps.
+        final BrokerValidator validator = new BrokerValidator(context);
+        try {
+            if (validator.verifySignature(inactiveBrokerPackageName)) {
```
Comment (sdabbiru) [inline common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java]: >inactiveBrokerPackageName [](start = 42, length = 25)

Will "inactiveBrokerPackage" be even present on the device at all times?

File: common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java [hunk 5]
```diff
@@ -140,4 +165,39 @@ private String getCurrentActiveBrokerPackageName(@NonNull final Context context)
...
+                return inactiveBrokerPackageName;
+            }
+        } catch (Exception e) {
+            // Log. It could be that inactiveBrokerPackageName is not installed on the app, or something's wrong.
```
Comment (sdabbiru) [inline common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java]: >app [](start = 87, length = 3)

device?

File: common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java [hunk 6]
```diff
@@ -140,4 +165,39 @@ private String getCurrentActiveBrokerPackageName(@NonNull final Context context)
...
+            Logger.error(TAG + methodName, "Failed to get inactive Broker package name.", e);
```
Comment (sdabbiru) [inline common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java]: >e [](start = 90, length = 1)

Will the exception e have sufficient information to distinguish between the 'inactive' broker package not installed on the device vs the package signature validation failure?

**Comment by rpdome** [comment]
Close as this should also support V1.

File: common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java [hunk 7]
```diff
@@ -140,4 +165,39 @@ private String getCurrentActiveBrokerPackageName(@NonNull final Context context)
...
+            return null;
+        }
+
+        // Verify the signature to make sure that we're not binding to malicious apps.
+        final BrokerValidator validator = new BrokerValidator(context);
+        try {
+            if (validator.verifySignature(inactiveBrokerPackageName)) {
+                return inactiveBrokerPackageName;
+            }
+        } catch (Exception e) {
```
Comment (sdabbiru) [inline common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java]: >Exception [](start = 17, length = 9)

I do not like this type of exception handling. I would like us to handle specific exceptions that are expected and then fallback on the 'Exception'.

Comment (rpdome) [reply common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java]: This one is closed. Please review https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/547 instead. Thanks!

**Pull request closed without merging**